dfanalyzer/graph.py

```diff
@@ -0,0 +1,195 @@
+            merge['interference'] = merge['min_dur']/merge['dur']
+            return merge
+
+        dft1, agg_dict, list_deg = dur_of_min_deg(self.ddf_deg)
+        self.inter = calculate_interference(dft1, agg_dict = agg_dict, list_deg = list_deg)
+    
+    
+    def write_checkpoint(self, id, cp_dir):
+        '''
+        write the datafame as parquet files
+        '''
+        write_df = getattr(self, id)
+        # print(write_df.compute())
+        # schema = {'id': int, 'name': str, 'pid': int, 'size': int, 'ts': int, 'te': int, 'mount_point': str, 'dur': int, 'trange': int, 'deg':int} 
+        write_df.to_parquet(f"{cp_dir}/{self.app_name}", name_function = lambda i:f'{id}-{i}.parquet')
+    
+
+    def read_checkpoint(self, id, cp_dir):
+        '''
+        read the dataframe from checkpoint files
+        '''
+        read_df = dd.read_parquet(f"{cp_dir}/{self.app_name}/{id}*.parquet")
+        setattr(self, id, read_df)
+
+
+class DFWorkflowgraph:
+    def __init__(self, ddf = None, app_name= "", trace_path = ''):
+        self.ddf = ddf
+        self.app_name = app_name
+        self.trace_path = trace_path
+
+    def select_cols(self,cols = []):
+        return self.ddf[cols]
+    
+    def get_pid_map(self):
+        '''
+        This function is designed for mummi traces to map pid with the application based on the filename
+        '''
+        all_files = glob.glob(self.trace_path)
+        pid_map = {}
+        for file in all_files:
+            slices = os.path.basename(file).split('.')
+            if (len(slices) > 4):
+                pid_map[slices[3]] = slices[1]
+        return pid_map
+
+
+
+    def create_workflow(self):
+        '''
+        1. Find number of times each file is prod/cons. And select the files that are both prod & cons atleast once
+        2. Get the list of the files which are both produced and consumed and select only the events with these files
+        '''
+        prod_cons = self.ddf.groupby('filename')['prod','cons'].sum()
+        prod_cons = prod_cons.query('prod > 0 and cons > 0').reset_index()
+        filelist = prod_cons.filename.unique().compute()
+        selected_events = self.ddf[self.ddf.filename.isin(filelist)]
+        selected_events_sum = selected_events.groupby(['filename', 'pid']).agg({'prod': 'sum', 'cons':'sum', 'ts':'min'}).reset_index()
+        merged = selected_events_sum.merge(prod_cons, on = ["filename"], how = 'left', suffixes = ['_pid','_fid'])
+        final = merged.query('not (prod_pid == prod_fid and cons_pid == cons_fid)')
+        return final
+    
+    def create_graph_df(self, df, pid_map):
+        '''
+        This needs to be moved to load cols.
+        '''
+        src_list = []
+        dest_list = []
+        wt_list = []
+        for index, row in df.iterrows():
```

Comment:
So typically, this can be written as a collection of Pandas queries. For instance, u can rename columns and use a mask to change values. I think this would be faster and better. If u have specific cases your unsure of we can work through that. 
as a rule of thumb never loop :D